pkg/scheduler/queue/tenant_queues_test.go

```diff
@@ -77,6 +79,65 @@ func TestQueues(t *testing.T) {
 	assert.NoError(t, err)
 }
 
+func TestQueuesRespectMaxTenantQueueSizeWithSubQueues(t *testing.T) {
+	maxTenantQueueSize := 100
+	qb := newQueueBroker(maxTenantQueueSize, 0)
+	additionalQueueDimensions := map[int][]string{
+		0: nil,
+		1: {"ingester"},
+		2: {"store-gateway"},
+		3: {"ingester-and-store-gateway"},
+	}
+	req := &SchedulerRequest{
+		Ctx:             context.Background(),
+		FrontendAddress: "http://query-frontend:8007",
+		UserID:          "tenant-1",
+		Request:         &httpgrpc.HTTPRequest{},
+	}
+
+	// build queue evenly with either no additional queue dimension or one of 3 additional dimensions
+	for i := 0; i < len(additionalQueueDimensions); i++ {
+		for j := 0; j < maxTenantQueueSize/len(additionalQueueDimensions); j++ {
+			req.AdditionalQueueDimensions = additionalQueueDimensions[i]
+			tenantReq := &tenantRequest{
+				tenantID: "tenant-1",
+				req:      req,
+			}
+			err := qb.enqueueRequestBack(tenantReq, 0)
+			assert.NoError(t, err)
+		}
+	}
+	// assert item count of tenant node and its subnodes
+	queuePath := QueuePath{"tenant-1"}
+	assert.Equal(t, maxTenantQueueSize, qb.tenantQueuesTree.getNode(queuePath).ItemCount())
+
+	// assert equal distribution of queue items between tenant node and 3 subnodes
+	for _, v := range additionalQueueDimensions {
+		queuePath := append(QueuePath{"tenant-1"}, v...)
+		assert.Equal(t, maxTenantQueueSize/len(additionalQueueDimensions), qb.tenantQueuesTree.getNode(queuePath).LocalQueueLen())
+	}
+
+	// assert error received when hitting a tenant's enqueue limit,
+	// even though most of the requests are in the subqueues
+	tenantReq := &tenantRequest{tenantID: "tenant-1", req: req}
+	err := qb.enqueueRequestBack(tenantReq, 0)
+	assert.ErrorIs(t, err, ErrTooManyRequests)
```

Comment:
Would be good to test both adding the request to the tenant node and adding the request to one of its subnodes